train.py

```diff
@@ -864,46 +881,65 @@ def train_one_epoch(
+        def _forward():
+            with amp_autocast():
+                output = model(input)
+                loss = loss_fn(output, target)
+            loss /= iters_to_accum
+            return loss
 
-        if not args.distributed:
-            losses_m.update(loss.item(), input.size(0))
-
-        optimizer.zero_grad()
-        if loss_scaler is not None:
-            loss_scaler(
-                loss, optimizer,
-                clip_grad=args.clip_grad,
-                clip_mode=args.clip_mode,
-                parameters=model_parameters(model, exclude_head='agc' in args.clip_mode),
-                create_graph=second_order
-            )
+        if need_step is not True and hasattr(model, "no_sync"):
+            with model.no_sync():
+                loss = _forward()
         else:
-            loss.backward(create_graph=second_order)
-            if args.clip_grad is not None:
-                utils.dispatch_clip_grad(
-                    model_parameters(model, exclude_head='agc' in args.clip_mode),
-                    value=args.clip_grad,
-                    mode=args.clip_mode
+            loss = _forward()
+
+        if not args.distributed:
+            losses_m.update(loss.item() * iters_to_accum, input.size(0))
+
+        def _backward():
+            if loss_scaler is not None:
+                loss_scaler(
+                    loss, optimizer,
+                    clip_grad=args.clip_grad,
+                    clip_mode=args.clip_mode,
+                    parameters=model_parameters(model, exclude_head='agc' in args.clip_mode),
+                    create_graph=second_order,
+                    need_step=need_step
                 )
-            optimizer.step()
+            else:
+                loss.backward(create_graph=second_order)
+                if args.clip_grad is not None:
+                    utils.dispatch_clip_grad(
+                        model_parameters(model, exclude_head='agc' in args.clip_mode),
+                        value=args.clip_grad,
+                        mode=args.clip_mode
+                    )
+                if need_step:
+                    optimizer.step()
 
-        if model_ema is not None:
-            model_ema.update(model)
+        if need_step is not True and hasattr(model, "no_sync"):
+            with model.no_sync():
+                _backward()
+        else:
+            _backward()
+            if need_step:
+                optimizer.zero_grad()
+                if model_ema is not None:
+                    model_ema.update(model)
 
-        torch.cuda.synchronize()
+                torch.cuda.synchronize()
+                num_updates += 1
+                batch_time_m.update(time.time() - end)
 
-        num_updates += 1
-        batch_time_m.update(time.time() - end)
-        if last_batch or batch_idx % args.log_interval == 0:
+        num_trains += input.size(0)
+        if need_step and (batch_idx // args.iters_to_accum) % args.log_interval == 0:
```

Comment:
something like `num_step_samples`  or `step_sample_count` might be more clear than `num_trains`

Reply:
I'll change the variable name as you mentioned!